File: LeetCode/Medium/0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o2[1] - o1[1]);
        
        for (int e : map.keySet()) {
            pq.offer(new int[] { e, map.get(e) });
        }

        int[] answer = new int[k];

        for (int i = 0; i < k; i++) {
            answer[i] = pq.poll()[0];
        }

        return answer;
    }
}

File: LeetCode/Medium/0347-top-k-frequent-elements/0347-top-k-frequent-elements.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val map = mutableMapOf<Int, Int>()

        for (num in nums) {
            map[num] = map.getOrDefault(num, 0) + 1
        }

        val pq = PriorityQueue<IntArray> { o1, o2 -> o2[1] - o1[1] }

        for (e in map.keys) {
            pq.offer(intArrayOf(e, map[e]!!))
        }

        val answer = IntArray(k)

        for (i in 0 until k) {
            answer[i] = pq.poll()[0]
        }

        return answer
    }
}